### ISSUE TITLE
Incorrect announce events

### ISSUE THREAD

**Issue by klimenttoshkov** [zerodahero/laravel-workflow#40]
Hi,

please review this [issue](https://github.com/symfony/symfony/issues/39981) with Symfony. Problem details are already explained there.

In review - they have inspected my bug report and could not reproduce it in vanilla Symfony install so their best guess is that the problem is in our event dispatcher. Could you please check?

**Comment by zerodahero** [comment]
Hi @klimenttoshkov , I'll try and take a closer look this week. At first glance, what the others are saying on that issue makes sense--seems like the guards would be the culprit here. There's very little that this package does in interaction with the Symfony component, when you run `workflow->apply(...)` that's being proxied directly to the Symfony components, and returned from there--events fired from there are fired from there, this package doesn't intervene at all--just wraps most with a Laravel-style event.

That said, it's probably worth taking a look at what the transition list look like immediately after transition. The announce event fires events for the same list returned by `$model->workflow_transitions()`

**Comment by klimenttoshkov** [comment]
Hi @zerodahero I know that, I have reviewed your Event subscriber and it is obvious for me that it just emits Laravel events on each Symfony event that is received.

I am also fully aware that a guard may be blocking a transition. I discovered the problem when I was testing what announce events are fired. When no guard is subscribed or if any exists, but it does not block the transition, I expect that transition to be announced but it is not.

It *may* be related to guards as what others guess, but I can't find how yet. The obvious possibility that a guard exists and it is blocking the transition is not the problem, you can inspect my sample app. Could you in fact test it and confirm your log?

**Comment by klimenttoshkov** [comment]
FYI I have tried adding guards for expected transitions and they are actually fired, which means they are checked as expected in order to fire appropriate announce events. Both guards return without blocking the transition, but the relevant announce event is not fired:

```
[2021-03-08 13:45:27] local.DEBUG: Event: workflow.test.announce  
[2021-03-08 13:45:27] local.DEBUG: Event: workflow.test.announce.toReady  
[2021-03-08 13:45:27] local.DEBUG: Event: workflow.test.guard  
[2021-03-08 13:45:27] local.DEBUG: Guard...  
[2021-03-08 13:45:27] local.DEBUG: Event: workflow.test.guard.tob1  
[2021-03-08 13:45:27] local.DEBUG: Event: workflow.test.guard  
[2021-03-08 13:45:27] local.DEBUG: Guard...  
[2021-03-08 13:45:27] local.DEBUG: Event: workflow.test.guard.tob2  
```

And as you can see, one announce event is actually fired,but it is for the transition that has completed (toReady).

At the same time output from Symfony is in correct events order and type:
```
❯ php src/app.php
Event name "workflow.guard"
Event name "workflow.acme.guard"
Event name "workflow.acme.guard.toReady"
Event name "workflow.leave"
Event name "workflow.acme.leave"
Event name "workflow.acme.leave.a"
Event name "workflow.transition"
Event name "workflow.acme.transition"
Event name "workflow.acme.transition.toReady"
Event name "workflow.enter"
Event name "workflow.acme.enter"
Event name "workflow.acme.enter.ready"
Event name "workflow.entered"
Event name "workflow.acme.entered"
Event name "workflow.acme.entered.ready"
Event name "workflow.completed"
Event name "workflow.acme.completed"
Event name "workflow.acme.completed.toReady"
Event name "workflow.announce"
Event name "workflow.acme.announce"
Event name "workflow.guard"
Event name "workflow.acme.guard"
Event name "workflow.acme.guard.tob1"
Event name "workflow.guard"
Event name "workflow.acme.guard"
Event name "workflow.acme.guard.tob2"
Event name "workflow.acme.announce.tob1"
Event name "workflow.acme.announce.tob2"
```

**Comment by klimenttoshkov** [comment]
@zerodahero  can you recreate this workflow in your own copy and test if correct announce events are fired?

**Comment by zerodahero** [comment]
@klimenttoshkov I'll try and make some time for that--it's been a busy week, but hopefully will get a chance by this weekend.

**Comment by zerodahero** [comment]
Hi, @klimenttoshkov sorry, this is still on my radar, it was a very busy week and weekend. I'm hoping to be able to take a look at it this week, though. Thanks for your patience!

**Comment by klimenttoshkov** [comment]
@zerodahero  you can try this [already prepared](https://github.com/klimenttoshkov/symfony-workflow-announce-event-test) test system.

**Comment by zerodahero** [comment]
@klimenttoshkov Ok, I found the issue. The workflow component (the Symfony part) only dispatches the specific events (`workflow.acme.announce.tob1`) and _not_ the primary event `workflow.announce`. We use the primary event to convert the Symfony event into a Laravel event and refire them all for the Laravel app.

Unfortunately, there's not a great solution given the current architecture. Symfony doesn't allow wildcard matching (`workflow.*.announce.*`) to subscribe to those events, and building them ahead of time based on the config is only an option if we remove the dynamic workflow aspect of this.

I'm looking into the possibility of removing this intermediary step of the event dispatching (Symfony -> package -> Laravel), and dispatch directly into the Laravel event system, which has a lot of potential. I'll keep you updated on that.

**Comment by klimenttoshkov** [comment]
@zerodahero I understand. Can you subscribe to all events and re-fire them? 

I have inspected your code in the event subscriber part and I understood that you re-fire all Symfony events.

And did this before opening the issue and was under the impression that you simply catch all Symfony events and raise the same Laravel events.

**Comment by zerodahero** [comment]
@klimenttoshkov I didn't see a way to subscribe to all events in Symfony. There's a few issues in the Symfony Events component asking for regex or similar, but they look like they were all closed with no changes. There's a wildcard event dispatcher package, but it doesn't support PHP 8.

**Comment by klimenttoshkov** [comment]
How do you subscribe now for the list of events if not using wildcard?

**Comment by klimenttoshkov** [comment]
Ok, I see this in `WorkflowSubscriber.php`:

```
    public function announceEvent(Event $event)
    {
        $workflowName = $event->getWorkflowName();
        $transitionName = $event->getTransition()->getName();

        $workflowEvent = new AnnounceEvent($event);

        event($workflowEvent);
        event('workflow.announce', $workflowEvent);
        event(sprintf('workflow.%s.announce', $workflowName), $workflowEvent);
        event(sprintf('workflow.%s.announce.%s', $workflowName, $transitionName), $workflowEvent);
    }

    public static function getSubscribedEvents()
    {
        return [
            'workflow.guard' => ['guardEvent'],
            'workflow.leave' => ['leaveEvent'],
            'workflow.transition' => ['transitionEvent'],
            'workflow.enter' => ['enterEvent'],
            'workflow.entered' => ['enteredEvent'],
            'workflow.completed' => ['completedEvent'],
            'workflow.announce' => ['announceEvent'],
        ];
    }
```

This looks to me that you subscribe to specific events and then when this event is raised, simply make up the rest of the events. Which is fine except for announce events because they cannot be predicted this way.

**Comment by klimenttoshkov** [comment]
I think we need to somehow wire Laravel event listener to Symfony event dispatcher?

**Comment by klimenttoshkov** [comment]
They suggested the following:

> One way forward for you is to use a custom implementation of the event dispatcher that includes the features you want, ie pattern matchning. That way you can register the events you want.

**Comment by zerodahero** [comment]
@klimenttoshkov Yep, that's my plan! Going to work on it today. I'll let you know how it goes.

**Comment by zerodahero** [comment]
Hi @klimenttoshkov , I've got a solution I think should work. I've pushed it up to develop. Give it a shot and let me know if it solves this!

**Comment by klimenttoshkov** [comment]
Current output from develop branch:

```
[2021-03-19 15:20:37] local.DEBUG: Event: workflow.test.entered
[2021-03-19 15:20:37] local.DEBUG: Event: workflow.test.entered.a
[2021-03-19 15:20:37] local.DEBUG: Event: workflow.test.guard
[2021-03-19 15:20:37] local.DEBUG: Event: workflow.test.guard.toReady
[2021-03-19 15:20:37] local.DEBUG: Event: workflow.test.leave
[2021-03-19 15:20:37] local.DEBUG: Event: workflow.test.leave.a
[2021-03-19 15:20:37] local.DEBUG: Event: workflow.test.transition
[2021-03-19 15:20:37] local.DEBUG: Event: workflow.test.transition.toReady
[2021-03-19 15:20:37] local.DEBUG: Event: workflow.test.enter
[2021-03-19 15:20:37] local.DEBUG: Event: workflow.test.enter.ready
[2021-03-19 15:20:37] local.DEBUG: Event: workflow.test.entered
[2021-03-19 15:20:37] local.DEBUG: Event: workflow.test.entered.ready
[2021-03-19 15:20:37] local.DEBUG: Event: workflow.test.completed
[2021-03-19 15:20:37] local.DEBUG: Event: workflow.test.completed.toReady
[2021-03-19 15:20:37] local.DEBUG: Event: workflow.test.announce
[2021-03-19 15:20:37] local.DEBUG: Event: workflow.test.guard
[2021-03-19 15:20:37] local.DEBUG: Guard...
[2021-03-19 15:20:37] local.DEBUG: Event: workflow.test.guard.tob1
[2021-03-19 15:20:37] local.DEBUG: Event: workflow.test.guard
[2021-03-19 15:20:37] local.DEBUG: Guard...
[2021-03-19 15:20:37] local.DEBUG: Event: workflow.test.guard.tob2
[2021-03-19 15:20:37] local.DEBUG: Event: workflow.test.announce.tob1
[2021-03-19 15:20:37] local.DEBUG: Event: workflow.test.announce.tob2
```

**Comment by zerodahero** [comment]
That looks correct now, yes?

**Comment by klimenttoshkov** [comment]
Yes, I believe.

**Comment by zerodahero** [comment]
Released in v3.3.0. Please let me know if you encounter anything else!
